File: solutions/problem-00012.scala
import scala.Math._

def divisors(n: Int) = 
  if (n == 1) 1 else 2 * (1 to sqrt(n)).filter(n % _ == 0).length

var n = 3
var i = 3
while (divisors(n) <= 500) {
  n += i
  i += 1
}

println(n)

File: haskell/problem-00012.hs
{-
[minhuang@mocha:haskell]$ time runghc problem-00012.hs
76576500

real  1m47.501s
user  1m46.503s
sys 0m0.040s
-}
tau :: Int -> Int
tau 1 = 1
tau n = 2 * x
        where sqn = floor . sqrt . fromIntegral $ n
              x = length $ filter (\i -> mod n i == 0) [1..sqn]

find n i | tau n > 500 = n
find n i | otherwise = find (n+i) (i+1)

result = find 3 3
main = print result
